### PR TITLE
[2/3] New firmware interfaces

### PR DESCRIPTION
**Pull request by joelrebel**
## What does this PR implement/change/remove?

Note: to be merged after https://github.com/bmc-toolbox/bmclib/pull/363

Defines new firmware interface methods - to enable individual firmware actions instead of a larger abstraction like the current `FirmwareInstall`, `FirmwareInstallStatus` - the abstraction in these current methods is getting to large to manage and work across vendors. 

`FirmwareUpload` - Only upload firmware
`FirmwareInstallUploaded` - Install previously uploaded firmware
`FirmwareTaskStatus` - Retrieve task state along with status information, to deprecate `FirmwareInstallStatus`
`FirmwareInstallSteps` - Returns steps the caller has to follow to install firmware for a device, since each vendor device is a special snowflake.

### Checklist
- [X] Tests added
- [X] Similar commits squashed

### CODE REVIEW

File: bmc/firmware.go [hunk 1]
```diff
@@ -5,25 +5,27 @@ import (
...
 
-// FirmwareInstaller defines an interface to install firmware updates
+// FirmwareInstaller defines an interface to upload and initiate a firmware install
 type FirmwareInstaller interface {
 	// FirmwareInstall uploads firmware update payload to the BMC returning the task ID
 	//
 	// parameters:
 	// component - the component slug for the component update being installed.
-	// applyAt - one of "Immediate", "OnReset".
+	// operationsApplyTime - one of the OperationApplyTime constants
```
Comment (DoctorVin) [inline bmc/firmware.go:22]: Why not make that a type instead of constants of `string`?

File: bmc/firmware.go [hunk 2]
```diff
@@ -165,3 +169,294 @@ func FirmwareInstallStatusFromInterfaces(ctx context.Context, installVersion, co
...
+}
+
+// firmwareInstallerProvider is an internal struct to correlate an implementation/provider and its name
+type firmwareInstallerWithOptionsProvider struct {
+	name string
+	FirmwareInstallerUploaded
+}
+
+// firmwareInstallUploaded uploads and initiates firmware update for the component
+func firmwareInstallUploaded(ctx context.Context, component, uploadTaskID string, generic []firmwareInstallerWithOptionsProvider) (installTaskID string, metadata Metadata, err error) {
```
Comment (DoctorVin) [inline bmc/firmware.go:193]: In general I think named returns are a maintenance hazard.

File: bmc/firmware.go (hunk 1 above)
Comment (joelrebel) [reply bmc/firmware.go:22]: Yep, that is the intention going ahead in the newer interface, for now we don't want to break the current interface.

File: bmc/firmware.go (hunk 2 above)
Comment (joelrebel) [reply bmc/firmware.go:193]: I kinda agree, although I have them there since they are indicative of what the returned string holds.

**Comment by joelrebel** [comment]
rebase on parent branch

**Pull request merged**